spec-0007/transition_to_rng.py

```diff
@@ -0,0 +1,156 @@
+                "2) an integer results in a different stream of random numbers, "
+                "(with the same distribution), and "
+                "3) `np.random` or `RandomState` instances result in an error. "
+                "See the documentation of `default_rng` for more information."
+            )
+
+            if as_old_kwarg:  # warn about deprecated use of old kwarg
+                kwargs[NEW_NAME] = kwargs.pop(old_name)
+                if dep_version:
+                    message = (
+                        f"Use of keyword argument `{old_name}` is "
+                        f"deprecated and replaced by `{NEW_NAME}`.  "
+                        f"Support for `{old_name}` will be removed "
+                        f"in SciPy {end_version}."
+                    ) + cmn_msg
+                    warnings.warn(message, DeprecationWarning, stacklevel=2)
+
+            elif as_pos_arg and dep_version:
+                # Warn about changing meaning of positional arg
+
+                # Note that this decorator does not deprecate positional use of the
+                # argument; it only warns that the behavior will change in the future.
+                # Simultaneously transitioning to keyword-only use is another option.
+
+                message = (
+                    f"Positional use of `{NEW_NAME}` (formerly known as "
+                    f"`{old_name}`) is still allowed, but the behavior is "
+                    "changing: the argument will be validated using "
+                    f"`np.random.default_rng` beginning in SciPy {end_version}."
+                ) + cmn_msg
+                warnings.warn(message, FutureWarning, stacklevel=2)
+
+            elif as_new_kwarg:  # no warnings; this is the preferred use
+                # After the removal of the decorator, validation with
+                # np.random.default_rng will be done inside the decorated function
+                kwargs[NEW_NAME] = np.random.default_rng(kwargs[NEW_NAME])
+
+            elif np.random.mtrand._rand._bit_generator._seed_seq is None:
+                # Emit FutureWarning if `np.random.seed` was used and no PRNG was passed
+                message = (
+                    "The NumPy global RNG was seeded by calling "
+                    f"`np.random.seed`. Beginning in {end_version}, this "
+                    "function will no longer use the global RNG."
+                ) + cmn_msg
+                warnings.warn(message, FutureWarning, stacklevel=2)
+
+            return fun(*args, **kwargs)
+
+        return wrapper
+
+    return decorator
+
+
+# Example usage of _prepare_rng decorator
+@_transition_to_rng("random_state", 1, "1.15.0")
+# previously, the signature of the function was
+#   library_function(arg1, random_state=None, arg2=None):
+def library_function(arg1, rng=None, arg2=None):
+    if isinstance(rng, (np.random.Generator, np.random.RandomState)):
+        rng = rng.random()
```

Comment:
IIRC, the ancient version tried to always normalize to `RandomState()` (also old input).

That seems nice.  It is a bit tedious to build a new args tuple if positional passing is allowed.  Alternatively, maybe expand this (and possibly the example in the main text) to be:
```
if rng is None:
    rng = np.random  # or np.random.mtrand._rand
elif not isinstance(rng, (np.random.Generator, np.random.RandomState)):
    rng = np.random.RandomState()
```
Or even just an:
```
if not isinstance(...):
    # old random state normalization code.
    pass
```

Reply:
> IIRC, the ancient version tried to always normalize to RandomState() (also old input).

In the older version of this code, normalization to `RandomState` was part of the decorator. I took that out because we want the decorator to leave input intended for the `random_state` argument alone and only work with input intended for the `rng` argument.

> Alternatively, maybe expand this (and possibly the example in the main text) to be:

I think I'm not understanding. This `library_function` is an example of an old function that alreay exists in the libary, and it is being wrapped to transition to the new `rng` keyword name and behavior. We assume that it already does what it wants to with input to `random_state` (we don't wish to change that behavior, whatever it is). We don't intend to recommend that the library make any changes inside the function itself except whatever is required to accept input with name `rng` instead of `random_state`. 

With that in mind, what is the intent of the suggested code?

Reply:
Maybe a typical function currently looks like:
```
random_state = normalize(random_state)
# use random_state
```
in which case, the typical new function will be:
```
if not isinstance(rng, (RandomState, Generator)):
   rng = normalize(rng)

# use rng the same was as random_state
```
I think I am missing the "unmodified" branch here in some form for clarity.  (Or just keep normalizing in the decorator, others can adapt that just as well.)

Anyway, not a big deal, I was mostly surprised by it.  It isn't hard to adapt after all, the question is only what provides the easier pattern to just take and incorporate.

Reply:
Ok I will respond to this by showing example versions of the library code before decorator, with decorator, and after decorator is removed.

Reply:
Addressed by https://github.com/stefanv/specs/pull/3/commits/cc43293281956828f2953e54952480de8d15d533